TESTS/usb_device/serial/main.cpp

```diff
@@ -0,0 +1,847 @@
+/*
+ * Copyright (c) 2018, ARM Limited, All Rights Reserved
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#if !defined(DEVICE_USBDEVICE) || !DEVICE_USBDEVICE
+#error [NOT_SUPPORTED] USB Device not supported for this target
+#endif
+
+#include "greentea-client/test_env.h"
+#include "utest/utest.h"
+#include "unity/unity.h"
+#include "mbed.h"
+#include <stdlib.h>
+#include "usb_phy_api.h"
+#include "USBCDC.h"
+#include "USBSerial.h"
+
+#define USB_CDC_VID 0x1f00
+#define USB_CDC_PID 0x2013
+#define USB_SERIAL_VID 0x1f00
+#define USB_SERIAL_PID 0x2012
+
+#define MSG_KEY_LEN 24
+#define MSG_VALUE_DUMMY "0"
+#define MSG_KEY_DEVICE_READY "ready"
+#define MSG_KEY_SERIAL_NUMBER "usb_dev_sn"
+#define MSG_KEY_PORT_OPEN_WAIT "port_open_wait"
+#define MSG_KEY_PORT_OPEN_CLOSE "port_open_close"
+#define MSG_KEY_SEND_BYTES_SINGLE "send_single"
+#define MSG_KEY_SEND_BYTES_MULTIPLE "send_multiple"
+#define MSG_KEY_LOOPBACK "loopback"
+#define MSG_KEY_CHANGE_LINE_CODING "change_lc"
+
+#define TX_BUFF_SIZE 32
+#define RX_BUFF_SIZE 32
+
+// The size of every data chunk the host sends (for each value from a
+// known sequence) during 'CDC RX multiple' test cases is
+// HOST_RX_BUFF_SIZE_RATIO times the size of RX_BUFF_SIZE input buffer.
+// This way the device has to correctly handle data bigger that its buffer.
+#define HOST_RX_BUFF_SIZE_RATIO 64
+
+// A DTR line is used to signal that the host has configured a terminal and
+// is ready to transmit and receive data from the USB CDC/Serial device.
+// When this test suite is run with the use of a Linux host, a workaround has
+// to be used to overcome some platform specific DTR line behavior.
+// Every time the serial port file descriptor is opened, the DTR line is
+// asserted until the terminal attributes are set.
+// As a consequence, the device receives a premature DTR signal with a
+// duration of 200-500 us before the correct, long-lasting DTR signal set by
+// the host-side test script. (tested on the Linux kernel 4.15.0)
+//
+// Online references:
+// https://github.com/pyserial/pyserial/issues/124#issuecomment-227235402
+//
+// The solution is to wait for the first DTR spike, ignore it, and wait for
+// the correct DTR signal again.
+#define LINUX_HOST_DTR_FIX 1
```

Comment:
Following a @donatieng suggestion, I checked how the USB gadget ACM driver might handle this case, and well.. it does not respect DTR state at all. :confused: 

https://github.com/torvalds/linux/blob/84df9525b0c27f3ebc2ebb1864fa62a97fdedb7d/drivers/usb/gadget/function/f_acm.c#L387-L392